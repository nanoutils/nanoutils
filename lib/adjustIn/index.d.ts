import {
  CurriedFunction1,
  CurriedFunction2,
  Fn,
  Predicate,
  MixedArray
} from '../../misc/types/fn'

declare function adjustIn<T1, T2>(fn: Fn<T1, T2>): CurriedFunction2<Predicate<T1>, Array<T1>, MixedArray<T1, T2>>
declare function adjustIn<T1, T2>(fn: Fn<T1, T2>, predicate: Predicate<T1>): CurriedFunction1<Array<T1>, MixedArray<T1, T2>>
declare function adjustIn<T1, T2>(fn: Fn<T1, T2>, predicate: Predicate<T1>, array: Array<T1>): MixedArray<T1, T2>

export default adjustIn