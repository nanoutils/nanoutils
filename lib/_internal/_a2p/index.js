export default (a) =>
  Array.isArray(a)
    ? a.join('.')
    : typeof a === 'string' || typeof a === 'number'
    ? '' + a
    : ''
