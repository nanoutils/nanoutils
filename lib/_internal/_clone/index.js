export function shallowCloneObjectAndArrays(obj) {
  var res = {}
  if (typeof obj !== 'object') {
    return res
  }
  if (Array.isArray(obj)) {
    return obj.slice()
  }
  for (var k in obj) {
    res[k] = obj[k]
  }
  return res
}
