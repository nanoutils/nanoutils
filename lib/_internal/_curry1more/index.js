export default function curry1more(f) {
  return function curried() {
    var args = [].slice.call(arguments)
    return args.length >= 2
      ? f.apply(f, args)
      : function() {
        return f.apply(f, args.concat([].slice.call(arguments)))
      }
  }
}
