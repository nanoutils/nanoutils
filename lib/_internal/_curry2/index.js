import placehold from '../_placehold'

export default function curry2(f) {
  return placehold(function curried(a, b) {
    return arguments.length >= 2
      ? f(a, b)
      : function(b2) {
        return f(a, b2)
      }
  }, 2)
}
