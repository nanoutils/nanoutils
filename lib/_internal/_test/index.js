/**
 * Check for accepting number of arguments for given function
 * @param {function} f - function to check
 * @param {number} n - number of accepting arguments
 * @param {array} args - example of arguments to check
 */
export function expectNumberOfArgs(f, n, args) {
  let k = n
  while (k > 0) {
    expect(typeof f).toBe('function')
    f = f(args[n - k--])
  }
  expect(typeof f).not.toBe('function')
}

export function expectExactNumberOfArgs(f, args) {
  expect(typeof f.apply(f, args)).not.toBe('function')
}
