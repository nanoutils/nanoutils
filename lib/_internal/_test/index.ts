import { expect } from "earljs";

/**
 * Check for accepting number of arguments for given function
 * @param {function} f - function to check
 * @param {number} n - number of accepting arguments
 * @param {array} args - example of arguments to check
 */
export function expectNumberOfArgs<Args extends unknown>(
  f: (args: Args) => unknown,
  n: number,
  args: Args[]
) {
  let k = n;
  let g: (args: Args) => unknown = f;
  while (k > 0) {
    expect(g).toBeA(Function);
    g = g(args[n - k--]) as (args: Args) => unknown;
  }
  expect(g).not.toBeA(Function);
}

export function expectExactNumberOfArgs<Args extends unknown>(
  f: (args: Args) => unknown,
  args: Args[]
) {
  expect(f.apply(f, args)).not.toBeA(Function);
}
