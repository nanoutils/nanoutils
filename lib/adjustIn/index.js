import curry3 from '../_internal/_curry3'
import adjust from '../adjust/adjust'
import findIndex from '../findIndex/findIndex'

export default curry3(function adjustIn(fn, predicate, array) {
  if (process.env.NODE_ENV !== 'production') {
    require('../_internal/_error').checkTypes('adjustIn', {
      fn: [fn, ['function'], typeof fn !== 'function'],
      predicate: [predicate, ['function'], typeof predicate !== 'function'],
      array: [array, ['array'], !Array.isArray(array)]
    })
  }
  return adjust(fn, findIndex(predicate, array), array)
})
