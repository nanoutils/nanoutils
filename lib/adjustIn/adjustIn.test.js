import { expectNumberOfArgs } from '../_internal/_test'
import adjustIn from '.'
import add from '../add'
import multiply from '../multiply'

const addTen = add(10)
const multiplyByTwo = multiply(2)

test('it accepts exact 3 arguments', () => {
  expectNumberOfArgs(
    adjustIn,
    3,
    [addTen, v => v === 1, [0, 1, 2]]
  )
})

test('applies function to array item with specified predicate', () => {
  expect(adjustIn(addTen, v => v === 4, [2, 3, 4, 5, 6])).toEqual([2, 3, 14, 5, 6])
  expect(adjustIn(multiplyByTwo, v => v === 6, [2, 3, 4, 5, 6])).toEqual([2, 3, 4, 5, 12])
})

test('return array itself if no element is found', () => {
  expect(adjustIn(addTen, v => v === 10, [2, 3, 4, 5, 6])).toEqual([2, 3, 4, 5, 6])
  expect(adjustIn(multiplyByTwo, v => v === 10, [2, 3, 4, 5, 6])).toEqual([2, 3, 4, 5, 6])
})

test('returns a copy even if no element is found', () => {
  const arr = [1, 2, 3]
  expect(adjustIn(addTen, v => v === 4, arr) === arr).toBeFalsy()
})
