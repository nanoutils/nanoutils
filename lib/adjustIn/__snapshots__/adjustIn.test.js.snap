// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`checks dev errors 1`] = `"[Nanoutils] adjustIn(fn, *, *) accepts a function, number given"`;

exports[`checks dev errors 2`] = `"[Nanoutils] adjustIn(*, predicate, *) accepts a function, boolean given"`;

exports[`checks dev errors 3`] = `"[Nanoutils] adjustIn(*, *, array) accepts an array, number given"`;
